### ISSUE TITLE
Understanding why contrastive loss is consistently better than triplet loss

### ISSUE THREAD

**Issue by w-hc** [KevinMusgrave/powerful-benchmarker#12]
Hi Kevin, thanks for the exhaustive work and stellar benchmarking.
I have 2 questions here.

1. Your results show that contrastive loss is consistently better than triplet loss. This runs against common (at least my personal) intuition that triplet loss demands less of the learned embedding space and makes learning easier. Would you mind suggesting some explanations as to why this intuition is incorrect? 

2. Are some losses easier to tune than others? Are some losses more robust to hyperparameter change / sampling strategy, and therefore easier to use in practice? My intuition suggests (maybe erroneously again) that a classification-based loss would have such properties (though temperature can be pesky too). This is not the focus of this work and is very hard to quantify, but some comments based on your personal user experience would be very valuable.

**Comment by KevinMusgrave** [comment]
Thanks for your interest!

1. Your intuition matches my intuition, as well as the explanations that are typically presented in papers. To be honest, I don't have a good explanation for why the triplet loss underperforms. Maybe an analysis of the gradients would lead to some insights. Maybe the training signal from the triplet loss is weak. For example, classification losses work very well and that seems like a much stronger signal (though I haven't actually derived the gradients so this is just a guess). I'm doing the benchmarking at a pretty high-level, i.e. I'm being hands-off about it, so for now I don't have much insight into the "why" of the results.

2. I haven't noticed a pattern in the speed of hyperparameter convergence yet. What I have noticed is that the best hyperparameters for classification losses (ArcFace, ProxyNCA etc) seem to lead to long training times, i.e. they keep improving bit by bit. It takes a long time for the accuracy to plateau.

By the way, I will eventually be putting all the results in the google spreadsheets, along with the config files, models, and bayesian optimization plots that show how the performance varies with hyperparameters. Those plots might lead to some insights.

**Comment by w-hc** [comment]
Thanks for the reply and looking forward to your works!

**Comment by KevinMusgrave** [comment]
@w-hc I'm starting to add stuff to [the spreadsheet](https://docs.google.com/spreadsheets/d/1brUBishNxmld-KLDAJewIc43A4EVZk3gY6yKe8OIKbY/edit#gid=0). 

You can click on the link in the "config files" column to see all the data. I'm planning to make a document where the optimization plots are more easily accessible, because right now most of them are html files which google drive doesn't preview properly.

For losses with just 1 hyperparameter, I added a separate HyperparameterPlot.png file, because I haven't tried getting Ax to display optimization plots for 1d. For losses with more than 2 hyperparameters, you can change the plot using the dropdown boxes.

**Comment by KevinMusgrave** [comment]
Just an FYI for anyone reading this: the numbers I put in the spreadsheet are going to be slightly different from the first version of the paper. Since writing the paper, I made improvements in logging to the benchmarker, so I've been re-running the experiments. I also made a slight change to how the validation plateaus are detected, and I also increased the weight decay from 5e-5 to 1e-4. I think that increase in weight decay may be reducing accuracy a bit.
